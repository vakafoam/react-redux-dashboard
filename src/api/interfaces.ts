export interface RequestParametersI {
  path: string;
  queryParams?: QueryParamsI;
  options?: object;
}

export interface QueryParamsI {
  keyword?: string; // search word
}

export interface PostRequestParametersI extends RequestParametersI {
  body?: object;
}

export interface APIErrorResponseI {
  status: number;
  path: string;
  titleCode: string;
  detailCode: string;
}

export interface PaginationI {
  size: number;
  totalElements: number;
  totalPages: number;
  number: number;
}

export interface MusicEventI {
  name: string;
  type: string;
  id: string;
  url: string;
  images: { ratio: string; url: string }[];
  dates: {
    start: { localDate: string; localTime: string };
    status: { code: string };
    timezone?: string;
  };
  promoter: object;
  _links: {
    self: {
      href: string;
    };
  };
  _embedded: {
    venues: {
      url: string;
      city: { name: string };
      country: { name: string; countryCode: string };
      address: { line1: string; line2?: string };
    }[];
  };
}

export interface MusicEventsResponseI {
  _embedded: {
    events: MusicEventI[];
  };
  _links: object;
  page: PaginationI;
}
