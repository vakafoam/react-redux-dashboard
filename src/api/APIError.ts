export default class APIError extends Error {
  status: number;
  ok: boolean;
  errorBody: any;

  constructor(response: any) {
    super(
      response.detail ||
        response.detailCode ||
        `API request ${response.path} failed with status ${response.status}`
    );
    this.status = response.status;
    this.ok = false;
    this.errorBody = response;
  }
}
