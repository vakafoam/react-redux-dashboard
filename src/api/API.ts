import APIError from "api/APIError";
import {
  APIErrorResponseI,
  PostRequestParametersI,
  RequestParametersI,
  QueryParamsI,
} from "api/interfaces";

const baseHeaders = { "Content-Type": "application/json; charset=utf-8" };

const parseResponseWithJson = async <ResultType>(
  response: Response
): Promise<ResultType> => {
  try {
    const result = await response.json();
    return result;
  } catch (error) {
    throw new APIError({
      status: response.status,
      path: response.url,
      titleCode: "errors.parsing.unparsableJson",
      detailCode: "errors.parsing.unparsableJson",
    });
  }
};

const getQueryParamsString = (
  apiKey?: string,
  params?: QueryParamsI
): string => {
  let queryString: string = "";
  if (params) {
    for (const [k, v] of Object.entries(params)) {
      queryString += `&${k}=${v}`;
    }
  }
  if (apiKey) {
    queryString += `&apikey=${apiKey}`;
  }
  return queryString;
};

export default class API {
  private apiKey = process.env.REACT_APP_API_KEY;

  async getJson<ResultType>({
    path,
    queryParams = {},
    options = {},
  }: RequestParametersI): Promise<
    { result: ResultType; headers: Headers } | undefined
  > {
    const fullPath = path + getQueryParamsString(this.apiKey, queryParams);

    const defaultOptions = {
      method: "GET",
      headers: {
        Accept: "application/json",
        "Accept-Language": "en",
      },
    };

    let response: Response;
    try {
      response = await fetch(fullPath, { ...defaultOptions, ...options });
      if (!response.ok) {
        const error: APIErrorResponseI = await parseResponseWithJson<
          APIErrorResponseI
        >(response);

        throw new APIError(error);
      }

      const result = await parseResponseWithJson<ResultType>(response);

      return {
        result,
        headers: response.headers,
      };
    } catch (e) {
      console.error("Error occured:", e);
    }
  }

  async postJson<ResultType>({
    path,
    body = {},
    queryParams = {},
    options = {},
  }: PostRequestParametersI) {
    return this.getJson<ResultType>({
      path,
      queryParams,
      options: {
        method: "POST",
        body: JSON.stringify(body),
        headers: baseHeaders,
        ...options,
      },
    });
  }
}
