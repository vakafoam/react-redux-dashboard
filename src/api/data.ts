import API from "api/API";
import { MusicEventsResponseI, QueryParamsI } from "api/interfaces";

export const MUSIC_EVENTS_URL =
  "https://app.ticketmaster.com/discovery/v2/events?countryCode=FI&classificationId=KZFzniwnSyZfZ7v7nJ";
export const MUSIC_GENRES_URL =
  "https://app.ticketmaster.com/discovery/v2/classifications/KZFzniwnSyZfZ7v7nJ.json";

export const getEventDetailsUrl = (id: string) =>
  `https://app.ticketmaster.com/discovery/v2/events/${id}`;

const api = new API();

export const getMusicEvents = (queryParams?: QueryParamsI) =>
  api.getJson<MusicEventsResponseI>({
    queryParams,
    options: {},
    path: MUSIC_EVENTS_URL,
  });

export const getClassificationsEvents = (queryParams?: QueryParamsI) =>
  api.getJson<any>({
    queryParams,
    options: {},
    path: MUSIC_GENRES_URL,
  });

export const getEventDetails = (id: string) =>
  api.getJson<any>({
    path: getEventDetailsUrl(id),
  });
