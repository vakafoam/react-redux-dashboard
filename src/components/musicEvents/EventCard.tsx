import React, { useState, SyntheticEvent } from "react";
import { MusicEventI } from "api/interfaces";
import EventDetails from "components/musicEvents/EventDetails";

interface PropsI {
  event: MusicEventI;
}

const get4x3Image = (event: MusicEventI) => {
  const image = event.images.find((i) => i.ratio === "4_3");
  return image ? image.url : event.images[0].url;
};

const EventCard = ({ event }: PropsI) => {
  const [isActive, setIsActive] = useState<boolean>(false);

  const onClick = (e: SyntheticEvent) => {
    e.stopPropagation();
    setIsActive(!isActive);
  };

  const onMouseEnter = () => setIsActive(true);
  const onMouseLeave = () => setIsActive(false);

  return (
    <div
      className="events__card"
      onClick={onClick}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
    >
      <img src={get4x3Image(event)} alt={event.name} />
      {isActive && <EventDetails event={event} />}
    </div>
  );
};

export default EventCard;
