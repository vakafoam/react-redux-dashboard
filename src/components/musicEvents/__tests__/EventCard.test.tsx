import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { MusicEventI } from "api/interfaces";
import EventCard from "components/musicEvents/EventCard";

const sampleMusicEvent: MusicEventI = {
  name: "Music event",
  type: "event",
  id: "12345",
  url: "htpp://example.com",
  images: [{ ratio: "4_3", url: "http://example2.com" }],
  dates: {
    start: { localDate: "some data", localTime: "some time" },
    status: { code: "code" },
  },
  promoter: {},
  _links: {
    self: {
      href: "http://goggle.com",
    },
  },
  _embedded: {
    venues: [
      {
        url: "https",
        city: { name: "Tallinn" },
        country: { name: "Estonia", countryCode: "EE" },
        address: { line1: "Tartu mnt" },
      },
    ],
  },
};

let container: Element | null = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  if (container) {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  }
});

describe("EventCard", () => {
  it("should render correctly and match snapshot", () => {
    act(() => {
      render(<EventCard event={sampleMusicEvent} />, container);
    });

    expect(container).toMatchSnapshot();
  });
});
