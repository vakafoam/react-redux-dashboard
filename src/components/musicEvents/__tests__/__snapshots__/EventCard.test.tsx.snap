// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventCard should render correctly and match snapshot 1`] = `
<div>
  <div
    class="events__card"
  >
    <img
      alt="Music event"
      src="http://example2.com"
    />
  </div>
</div>
`;
